src/bb_excel/core.clj

```diff
@@ -130,127 +100,178 @@
+         ^long sheet-id (get-and-check-sheet-id zipfile sheetname-or-idx)
+         ^ZipEntry sheet-entry (get-sheet-entry zipfile sheet-id)]
      (let [opts    (merge defaults options)
            row     (:row opts)
            hdr     (:hdr opts)
            row     (if (and hdr (zero? row)) 1 row)
            rows    (:rows opts)
            fxn     (:fxn opts)
            cols    (map fxn (:columns opts))
-           sheetid (cond
-                     (string? sheetname-or-idx)
-                     (:idx (first (filter #(= sheetname-or-idx (:name %)) (get-sheet-names file-or-filename))))
-
-                     (and (integer? sheetname-or-idx) (pos? sheetname-or-idx))
-                     sheetname-or-idx
-
-                     :else
-                     (let [message (format "Attr 'sheetname-or-idx' can only be string or positive number, but passed '%s'" sheetname-or-idx)]
-                       (throw (ex-info message {}))))
-           wb      (.getEntry zipfile (str "xl/worksheets/sheet" sheetid ".xml"))
-           ins     (.getInputStream zipfile wb)
            shared-strings (get-shared-strings zipfile)
-           styles  (get-styles zipfile)
-           xx      (slurp ins)
-           x       (parse-str xx)
-           d       (->>  (:content x)
-                         (filter #(contains? SHEET_DATA_TAGS (:tag %)))
-                         first :content
-                         (map :content)
-                         (take rows)
-                         (map (partial process-row shared-strings styles)));
-           dx (remove #(= row (:_r %)) d)
-           h (when hdr (merge (update-vals (first (filter #(= (:_r %) row) d)) fxn) {:_r :_r}))
-           dy (if (pos? rows)
-                (take rows (map #(rename-keys % h) dx))
-                (map #(rename-keys % h) dx))]
-       (if (empty? cols) dy (map #(select-keys % cols) dy)))
-     (let [message (format "Attr 'file-or-filename' contains value not suitable for creating ZipFile: '%s'" file-or-filename)]
-       (throw (ex-info message {}))))))
+           styles  (get-styles zipfile)]
+       (with-open [sheet (.getInputStream zipfile sheet-entry)]
+         (let [worksheet-node (xml/parse sheet {:namespace-aware false})
+               sheet-data-node (->> (:content worksheet-node)
+                                    (find-first (by-tag :sheetData)))
+               row-nodes (:content sheet-data-node)
+               d (->> row-nodes
+                      (take rows)
+                      (process-rows shared-strings
+                                    styles
+                                    BASE_ROW_INDEX))
+               dx (remove #(= row (:_r %)) d)
+               h (when hdr (merge (update-vals (first (filter #(= (:_r %) row) d)) fxn)
+                                  {:_r :_r}))
+               dy (if (pos? rows)
+                    (take rows (mapv #(rename-keys % h) dx))
+                    (mapv #(rename-keys % h) dx))]
```

Comment:
If you find time please explain if we drop some rows from begining and add hdr option. Header column name should take from top of whole file(row with index 1) or from top of remaining rows(row with row-opt index)?

Reply:
Excel does the latter if i'm not mistaken.  So we use top of remaining rows as header column.